File: apps/api/src/routes/apiKeys.ts
import { Router, Response } from 'express';
import { AuthRequest, authenticateRequest } from '../middleware/auth';
import { createApiKey, deleteApiKey, listApiKeys, writeAuditLog } from '../database';
import { generateApiKey, hashApiKey } from '../services/utils';

const router = Router();
router.use(authenticateRequest);

function clientIp(req: AuthRequest): string | null {
  const forwarded = req.headers['x-forwarded-for'];
  if (typeof forwarded === 'string' && forwarded.trim()) return forwarded.split(',')[0].trim();
  return req.ip || null;
}

router.get('/', async (req: AuthRequest, res: Response) => {
  if (!req.tenantId) return res.status(401).json({ error: 'Unauthorized' });
  const keys = await listApiKeys(req.tenantId);
  return res.json({ apiKeys: keys });
});

router.post('/', async (req: AuthRequest, res: Response) => {
  if (!req.tenantId || !req.userId) return res.status(401).json({ error: 'Unauthorized' });

  const { name, scopes, expiresAt } = req.body as {
    name?: string;
    scopes?: string[];
    expiresAt?: string;
  };

  const keyName = (name || 'CLI key').trim();
  if (!keyName) return res.status(400).json({ error: 'Name is required' });

  const parsedScopes = Array.isArray(scopes) && scopes.length
    ? scopes.map(String)
    : ['deploy', 'read'];

  let expiry: Date | null = null;
  if (expiresAt) {
    expiry = new Date(expiresAt);
    if (Number.isNaN(expiry.getTime())) {
      return res.status(400).json({ error: 'expiresAt must be an ISO date' });
    }
  }

  const { key, prefix } = generateApiKey();
  const record = await createApiKey({
    tenantId: req.tenantId,
    userId: req.userId,
    name: keyName,
    keyHash: hashApiKey(key),
    prefix,
    scopes: parsedScopes,
    expiresAt: expiry,
  });

  await writeAuditLog({
    tenantId: req.tenantId,
    userId: req.userId,
    action: 'api_key.create',
    resourceType: 'api_key',
    resourceId: record.id,
    changes: { name: keyName, scopes: parsedScopes },
    ipAddress: clientIp(req),
  });

  return res.status(201).json({ apiKey: record, key });
});

router.delete('/:id', async (req: AuthRequest, res: Response) => {
  if (!req.tenantId || !req.userId) return res.status(401).json({ error: 'Unauthorized' });

  const ok = await deleteApiKey(req.params.id, req.tenantId);
  if (!ok) return res.status(404).json({ error: 'API key not found' });

  await writeAuditLog({
    tenantId: req.tenantId,
    userId: req.userId,
    action: 'api_key.delete',
    resourceType: 'api_key',
    resourceId: req.params.id,
    ipAddress: clientIp(req),
  });

  return res.status(204).send();
});

export default router;

File: apps/api_python/routes/api_keys.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Request, Response, status

import database as db
from auth import AuthContext, authenticate_request, client_ip
from schemas import ApiKeyCreate
from services.utils import generate_api_key, hash_api_key

router = APIRouter()


@router.get('/')
async def list_keys(auth: AuthContext = Depends(authenticate_request)):
    return {'apiKeys': db.list_api_keys(auth.tenant_id)}


@router.post('/', status_code=status.HTTP_201_CREATED)
async def create_key(payload: ApiKeyCreate, request: Request, auth: AuthContext = Depends(authenticate_request)):
    if not auth.user_id:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail='API key has no user; sign in to create keys')

    key_name = (payload.name or 'CLI key').strip()
    if not key_name:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail='Name is required')

    scopes = payload.scopes if payload.scopes else ['deploy', 'read']
    expiry = None
    if payload.expiresAt:
        expiry = datetime.fromisoformat(payload.expiresAt.replace('Z', '+00:00'))
        if expiry.tzinfo is None:
            expiry = expiry.replace(tzinfo=timezone.utc)

    key, prefix = generate_api_key()
    record = db.create_api_key({
        'tenantId': auth.tenant_id,
        'userId': auth.user_id,
        'name': key_name,
        'keyHash': hash_api_key(key),
        'prefix': prefix,
        'scopes': scopes,
        'expiresAt': expiry,
    })

    db.write_audit_log({
        'tenantId': auth.tenant_id,
        'userId': auth.user_id,
        'action': 'api_key.create',
        'resourceType': 'api_key',
        'resourceId': record['id'],
        'changes': {'name': key_name, 'scopes': scopes},
        'ipAddress': client_ip(request),
    })

    return {'apiKey': record, 'key': key}


@router.delete('/{key_id}', status_code=status.HTTP_204_NO_CONTENT)
async def delete_key(key_id: str, request: Request, auth: AuthContext = Depends(authenticate_request)):
    if not auth.user_id:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail='Unauthorized')

    if not db.delete_api_key(key_id, auth.tenant_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail='API key not found')

    db.write_audit_log({
        'tenantId': auth.tenant_id,
        'userId': auth.user_id,
        'action': 'api_key.delete',
        'resourceType': 'api_key',
        'resourceId': key_id,
        'ipAddress': client_ip(request),
    })
    return Response(status_code=status.HTTP_204_NO_CONTENT)
